extensions/hibernate-search-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/elasticsearch/runtime/HibernateSearchElasticsearchRuntimeConfig.java

```diff
@@ -455,4 +461,30 @@ public String getHibernateSearchString() {
         @ConfigItem(defaultValueDocumentation = "10S")
         Optional<Duration> requiredStatusWaitTimeout;
     }
+
+    // We can't set actual default values in this section,
+    // otherwise "quarkus.hibernate-search.elasticsearch.index-defaults" will be ignored.
+    @ConfigGroup
+    public static class ElasticsearchIndexIndexingConfig {
+        /**
+         * The number of indexing queues assigned to each index.
+         */
+        // We can't set an actual default value here: see comment on this class.
+        @ConfigItem(defaultValueDocumentation = "10")
```

Comment:
I'm not entirely sure what a reasonable default would be here - so bear with me as I only know the high level picture.

I wonder if this is a reasonable value for an high-performance index. But maybe we should default to small and lean for a Quarkus default?

Reply:
Just so you know, queues don't have an assigned thread. Threads will only be used when we process the queues and to orchestrate the works, but after the first work is sent, everything happens in the HTTP client threads.

So you can raise this to 256, if you have a thread pool of size 8, you'll still have only 8 threads. Since queue processing is non-blocking, that shouldn't be a problem.

The number of queues defines the level of parallelism, i.e. with 10 queues you will have 10 indexing requests sent in parallel. Crucially, this means we'll use at most 10 connections in parallel for indexing for this index.

The "best" value will likely depend on the size of your cluster and the number of available connections. I haven't been able to pinpoint a "best" value here, as the setting will have a different effect depending on the number of indexes being written to in parallel, the number of available connections, the topology of your Elasticsearch cluster, ...

Bottom line: I don't know what's best for the user either. Which is why I expose this configuration option: so that they can set it themselves. I can change the default to 5, but that'd be as much guesswork as it is now.

Maybe after some more performance testing we'll get a better idea of what the default should be.

Reply:
Thanks for the explanation.  Yes 10 might be a good starting point then but it feels like a random number - since you mentioned they are non-blocking perhaps the best default is 1? One connection should be able to make a quite efficient use of the available network resources - similar systems will use one by default, and maybe 2 or 3 when they really need to push for the maximum.

This also makes me wonder why we expose this. Would you agree that what matters for end users is to understand that this is about the number of connections? Maybe it should be named and documented like that.

Reply:
(again: when in doubt or when you believe it won't matter much, make a choice and don't expose it)

Reply:
**TL;DR**: we don't have great defaults for the most common setups yet, and we will never have defaults that fit every setup. Hence the newly exposed settings.

> One connection should be able to make a quite efficient use of the available network resources - similar systems will use one by default, and maybe 2 or 3 when they really need to push for the maximum.

I'm not sure, but you have more experience about network performance than I do. If you're certain about that, I will definitely change this setting.

From what I've seen with Elasticsearch... in raw throughput tests, where we just push everything we can to the index, and there is only one Elasticsearch instance, you're absolutely right. Less connections is even better, probably because it means larger bulks and those are better handled by Elasticsearch.

However, when I spin up a complete application with database access, with 2 (local) Elasticsearch nodes, and I start mass indexing, I witnessed the opposite: jumping from 2 queues to 10 queues gave a +50% throughput increase in a small local test (~25000 wikipedia pages) I performed. Granted there's nothing scientific to that test, but it's at least enough to give pause about whether "more connections is pointless".

There are other things that come into play, I suppose; maybe optimizing the global indexing time while allowing high latency from time to time for some documents is not what works best in all situations.

> This also makes me wonder why we expose this.

Because I can't come up with defaults that will work for the most common setup (which is it, by the way?), much less for everyone. At least, not yet.

> when in doubt or when you believe it won't matter much, make a choice and don't expose it

I don't believe it won't matter, I just believe I don't have a way to come up with the best set of settings for everyone, especially with varying environments: local Elasticsearch instance vs remote, single ES instance vs 100+ nodes, 1 index vs 50, ...

I can postpone exposing these settings until we spend a few weeks doing reliable, scientific performance tests, but who knows when that will happen... At least if I expose these settings, users have a way to improve performance while we work on implementing missing features. And who knows, maybe someone tinkering with these settings will give us useful feedback :)

Reply:
That's great that you tested it. Ok then keep it :)

> And who knows, maybe someone tinkering with these settings will give us useful feedback :)

+1 that's a very good point. And a good reason to document some of your suggestions, so that others power users can actually do that.